### PR TITLE
[BugFix] Fix lake update manager graceful shutdown problem

### PR DESCRIPTION
**Pull request by TszKitLo40**
Why I'm doing:
`_compaction_state_mem_tracker` will be destructed before `_compaction_cache`. When `_compaction_cache` is destructed, `_compaction_state_mem_tracker` maybe nullptr.
What I'm doing:
Make `_compaction_state_mem_tracker` be destructed after `_compaction_cache`.
Fixes #issue

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/storage/lake/update_manager.h [hunk 1]
```diff
@@ -195,9 +193,12 @@ class UpdateManager {
     std::unique_ptr<MemTracker> _update_state_mem_tracker;
     std::unique_ptr<MemTracker> _compaction_state_mem_tracker;
 
+    // compaction cache
+    DynamicCache<string, CompactionState> _compaction_cache;
```
Comment (kevincai) [inline be/src/storage/lake/update_manager.h:197]: add destructor to control the member destruction order.

**Review by kevincai — APPROVED** [review]
better add a simple UT to cover the case.

File: be/src/storage/lake/update_manager.cpp [hunk 2]
```diff
@@ -43,12 +43,17 @@ UpdateManager::UpdateManager(LocationProvider* location_provider, MemTracker* me
...
     _update_state_cache.set_mem_tracker(_update_state_mem_tracker.get());
+    _compaction_cache.set_mem_tracker(_compaction_state_mem_tracker.get());
 
     int64_t byte_limits = ParseUtil::parse_mem_spec(config::mem_limit, MemInfo::physical_mem());
     int32_t update_mem_percent = std::max(std::min(100, config::update_memory_limit_percent), 0);
     _index_cache.set_capacity(byte_limits * update_mem_percent);
 }
 
+UpdateManager::~UpdateManager() {
+    _compaction_cache.clear();
```
Comment (luohaha) [inline be/src/storage/lake/update_manager.cpp:54]: You can also clear `_index_cache` and `_update_state_cache`

File: be/src/storage/lake/update_manager.cpp [hunk 3]
```diff
@@ -43,12 +43,17 @@ UpdateManager::UpdateManager(LocationProvider* location_provider, MemTracker* me
     _compaction_state_mem_tracker = std::make_unique<MemTracker>(-1, "compaction_state_cache", mem_tracker);
     _index_cache.set_mem_tracker(_index_cache_mem_tracker.get());
     _update_state_cache.set_mem_tracker(_update_state_mem_tracker.get());
+    _compaction_cache.set_mem_tracker(_compaction_state_mem_tracker.get());
```
Comment (luohaha) [inline be/src/storage/lake/update_manager.cpp:46]: I think we don't need this here.

**Review by wyb — APPROVED** [review]

**Review by sduzh — APPROVED** [review]

**Pull request merged**